### PR TITLE
C#: Fix attribute argument extraction

### PR DESCRIPTION
**Pull request by tamasvajk**
In some cases matching the attribute argument syntax to the symbol was throwing an exception:
```
[ERROR]   Uncaught exception in Specified argument was out of the range of valid values. (Parameter 'index')    at Microsoft.CodeAnalysis.SeparatedSyntaxList`1.get_Item(Int32 index)
   at Semmle.Extraction.CSharp.Entities.Attribute.ExtractArguments(TextWriter trapFile) in /home/jenkins/workspace/ODASA/job/code/ql/csharp/extractor/Semmle.Extraction.CSharp/Entities/Attribute.cs:line 72
   at Semmle.Extraction.CSharp.Entities.Attribute.Populate(TextWriter trapFile) in /home/jenkins/workspace/ODASA/job/code/ql/csharp/extractor/Semmle.Extraction.CSharp/Entities/Attribute.cs:line 65
   at Semmle.Extraction.Context.<>c__DisplayClass52_0.<Populate>b__4() in /home/jenkins/workspace/ODASA/job/code/ql/csharp/extractor/Semmle.Extraction/Context.cs:line 373
   at Semmle.Extraction.ContextExtensions.Try(Context context, SyntaxNode node, ISymbol symbol, Action a) in /home/jenkins/workspace/ODASA/job/code/ql/csharp/extractor/Semmle.Extraction/Context.cs:line 518
```

@hvitved I'm unsure how I could have missed these in the differences job. I assume that these were not reported there, but why? Also, we have shipped this bug with 1.26, what should we do about it?

We were looking up the `syntax` by index, and if the parameter had a default value, then we were indexing outside the bounds, as there was no syntax for the symbol.

In a second commit, I also fixed argument lookup in case an argument name is present, such as in `a: true`. In this case we're looking up the syntax by the name, and not by index.

~~[C# differences job](https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/728/)~~
[C# differences job](https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/736/)

### CODE REVIEW

**Review by hvitved — COMMENTED** [review]
The first commit LGTM, but I am not convinced about the second commit. For example, if you change
```csharp
public My2Attribute(bool a, bool b, int i = 12) { }
```
to
```csharp
public My2Attribute(bool a, bool b, int i = 12, int j = 13) { }
```
and
```csharp
[My2(b: true, a: false, X = 42)]
```
to
```csharp
[My2(b: true, j: 1, a: false, X = 42)]
```
then the test output looks incorrect to me.

For normal calls, we handle named arguments in QL and not in the extractor, so I think we should do the same for the attribute constructor calls.

**Comment by tamasvajk** [comment]
@hvitved Thanks for checking. I'll fix the second commit. I'm not sure if we can do the same here as we do with normal method calls. `symbol.ConstructorArguments` doesn't give us the argument name, it only returns the reflected argument data.

File: csharp/extractor/Semmle.Extraction/Entities/Location.cs [hunk 1]
```diff
@@ -13,6 +15,14 @@ protected Location(Context cx, Microsoft.CodeAnalysis.Location? init)
                     ? NonGeneratedSourceLocation.Create(cx, loc)
                     : Assembly.Create(cx, loc);
 
+        public static Location Create(Context cx)
+        {
+            return cx.SourceTree == null
+                ? GeneratedLocation.Create(cx)
+                : Create(cx, Microsoft.CodeAnalysis.Location.Create(cx.SourceTree, TextSpan.FromBounds(0, 0)));
+
```
Comment (tamasvajk) [inline csharp/extractor/Semmle.Extraction/Entities/Location.cs:23]: Remove empty line.

**Comment by tamasvajk** [comment]
@hvitved I pushed an extra commit to this PR as a first commit to use a file based error message location when possible instead of the generated location. I started the differences job again: https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/736/

**Comment by tamasvajk** [comment]
> @hvitved I pushed an extra commit to this PR as a first commit to use a file based error message location when possible instead of the generated location. I started the differences job again: https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/736/

@hvitved The differences job shows the removed errors. But we still have no real location in the UI. It says `Location information not available.`, which comes from [here](https://github.com/github/semmle-code/blob/main/buildutils-internal/jenkins/Changes/summarise_changes.py#L123). The `.sarif` file has a slightly different format for these 0 length spans:
```
"locations" : [ {
  "physicalLocation" : {
    "artifactLocation" : {
      "uri" : ...,
      "uriBaseId" : "%SRCROOT%",
      "index" : 1542
    },
    "region" : {
      "charOffset" : 0,
      "charLength" : 0
    }
  }
```
vs the expected
```
"region" : {
  "startLine" : ...,
  "startColumn" : ...,
  "endLine" : ...,
  "endColumn" : ...
}
```

I think it's good enough. I checked in QL, these 0-0 locations can be queried.

**Comment by hvitved** [comment]
> @hvitved The differences job shows the removed errors.

Awesome, thanks for making this change. Let's get it in!

**Review by hvitved — APPROVED** [review]

**Review by hvitved — APPROVED** [review]

**Pull request merged**